File: solutions/clojure/log-levels/5/src/log_levels.clj
(ns log-levels
  (:require [clojure.string :as str]))

(defn message
  "Takes a string representing a log line
   and returns its message with whitespace trimmed."
  [log-line]
  (str/trim (last (str/split log-line #":"))))

(defn log-level
  "Takes a string representing a log line
   and returns its level in lower-case."
  [log-line]
  (str/lower-case (last (re-find #"\[([A-Z]+)\]" log-line))))

(defn reformat
  "Takes a string representing a log line and formats it
   with the message first and the log level in parentheses."
  [log-line]
  (let [new-log-format "%s (%s)"]
  (format new-log-format (message log-line) (log-level log-line))))

File: solutions/cpp/log-levels/2/log_levels.cpp
#include <string>

namespace log_line {
    // Log line format: `[LEVEL]: MESSAGE`
    std::string message(std::string line) {
        int msg_start_idx = line.find(":") + 2;
        return line.substr(msg_start_idx);
    }

    std::string log_level(std::string line) {
        int level_start_idx = line.find("[") + 1;
        int level_length = line.find("]") - level_start_idx;
        return line.substr(level_start_idx, level_length);
    }
    
    std::string reformat(std::string line) {
        return message(line) + " (" + log_level(line) + ")";
    }
}  // namespace log_line
